kubetest/kops.go

```diff
@@ -487,3 +530,89 @@ func (k *kops) runKopsDump() (*kopsDump, error) {
 
 	return dump, nil
 }
+
+// getRandomAWSZones looks up all regions, and the availability zones for those regions.  A random
+// region is then chosen and the AZ's for that region is returned.  If HA is required a regions with 3
+// or more zones is returned.
+func getRandomAWSZones(masterCount int) (*[]string, error) {
+	ec2Session, err := getAWSEC2Session("us-east-2")
+
+	regionResults, err := ec2Session.DescribeRegions(&ec2.DescribeRegionsInput{})
+	if err != nil {
+		return nil, fmt.Errorf("unable to call aws api DescribeRegions: %v", err)
+	}
+
+	var regions []string
+	regionMap := make(map[string][]string)
+
+	// we have to iterate over the regions to get the AZs for each region.
+	for _, r := range regionResults.Regions {
+		regions = append(regions, *r.RegionName)
+		ec2Session, err := getAWSEC2Session(*r.RegionName)
+
+		// az for each of the regions. AWS Go API does not allow us to make a single call
+		zonesResults, err := ec2Session.DescribeAvailabilityZones(&ec2.DescribeAvailabilityZonesInput{})
+		if err != nil {
+			return nil, fmt.Errorf("unable to call aws api DescribeAvailabilityZones: %v", err)
+		}
+		// build a map of region and the regions az so that we can lookup AZ easily
+		for _, z := range zonesResults.AvailabilityZones {
+			existing, ok := regionMap[*z.RegionName]
+			var az []string
+			if ok {
+				az = append(existing, *z.ZoneName)
+			} else {
+				az = []string{*z.ZoneName}
+			}
+			regionMap[*z.RegionName] = az
+		}
+	}
+
+	// initialize global pseudo random generator
+	rand.Seed(time.Now().Unix())
+
+	// select a random region and zones
```

Comment:
You asked for more context on using rand.Perm.

Instead of looping through and removing the regions one by one, instead just use `rand.Perm` to get a list of indexes.  Those are the random order in which you will consider the regions in the slice.  Then you don't need to do the removal, and you only need one rand call.

```
for _, i := rand.Perm(n) {
  region := regions[i]
...
}
return nil, fmt.Errorf("unable to find region with %d zones", masterCount)
```